File: Creational/Lab2/Python/agency.py
import random
import threading
class Agency:
    _instance = None
    _lock_instance = threading.Lock()
    _lock_apartment = threading.Lock()
    _lock_room = threading.Lock()
    _lock_house = threading.Lock()
    _lock_realtor = threading.Lock()

    def __init__(self):
        self.apartments = []
        self.rooms = []
        self.houses = []
        self.realtors = []

    @classmethod
    def get_instance(cls):
        if cls._instance is None:
            with cls._lock_instance:
                if cls._instance is None:
                    cls._instance = cls()
        return cls._instance

    def add_apartment(self, apartment):
        with self._lock_apartment:
            self.apartments.append(apartment)

    def add_room(self, room):
        with self._lock_room:
            self.rooms.append(room)

    def add_house(self, house):
        with self._lock_house:
            self.houses.append(house)

    def add_realtor(self, realtor):
        with self._lock_realtor:
            self.realtors.append(realtor)

    def print_random_estate_and_clones(self):
        r = random.Random()
        i = r.randint(0, 2)
        estate = []

        if i == 0:
            estate = self.apartments
        elif i == 1:
            estate = self.rooms
        elif i == 2:
            estate = self.houses

        if len(estate) > 0:
            j = r.randint(0, len(estate) - 1)
            el = estate[j]
            el_clone = el.clone()
            el_copy = el.shallow_copy()

            print(f"Original: {el}\n")
            print(f"Clone: {el_clone}\n")
            print(f"Shallow copy: {el_copy}\n")

            print("Change shallow copy realtor name to 'Jane Doe'...\n")
            el_copy.realtor.name_and_surname = "Jane Doe"
            print(f"Original after changing shallow copy: {el}\n")
            print(f"Clone: {el_clone}")

File: Creational/Lab2/C#/Agency.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace C_
{
    public class Agency
    {
        private Agency() {
            _apartments = new List<Apartment>();
            _realtors = new List<Realtor>();
            _houses = new List<House>();
            _rooms = new List<Room>();
        }

        //Instances
        private static Agency _instance;
        private static List<Apartment> _apartments;
        private static List<Room> _rooms;
        private static List<House> _houses;
        private static List<Realtor> _realtors;

        //Locks
        private static readonly object _lockInstance = new object();
        private static readonly object _lockApartments = new object();
        private static readonly object _lockRooms = new object();
        private static readonly object _lockHouses = new object();
        private static readonly object _lockRealtors = new object();

        public static Agency GetInstance()
        {
            if (_instance == null)
            {
                lock (_lockInstance)
                {
                    if (_instance == null)
                        _instance = new Agency();
                }
            }
            return _instance;
        }

        public  List<Apartment> Apartments 
        {
            get 
            {
                lock (_lockApartments)
                {
                    return _apartments;
                }
            }
        }
        public List<Room> Rooms
        {
            get 
            {
                lock (_lockRooms)
                {
                    return _rooms;
                }
            }
        }
        public List <House> Houses
        {
            get 
            {
                lock (_lockHouses)
                {
                    return _houses;
                }
            }
        }
        public List<Realtor> Realtors 
        { 
            get 
            {
                lock (_lockRealtors)
                {
                    return _realtors;
                }
            }
        }

        public void AddRealtor(Realtor realtor){
            lock (_lockRealtors) {
                _realtors.Add(realtor);
            }
        }
        public void RemoveRealtor(Realtor realtor){
            lock(_lockRealtors) {
                _realtors.Remove(realtor);
            }
        }

        public void AddEstate(RealEstate estate){
            switch (estate.GetType().Name) {
                case "Apartment":
                    lock(_lockApartments){
                        _apartments.Add((Apartment)estate);
                    };
                    break;
                case "Room":
                    lock(_lockRooms){
                        _rooms.Add((Room)estate);
                    };
                    break;
                case "House":
                    lock(_lockHouses){
                        _houses.Add((House)estate);
                    };
                    break;
            }
        }

        public void RemoveEstate(RealEstate estate){
            switch (estate.GetType().Name) {
                case "Apartment":
                    lock(_lockApartments){
                        _apartments.Remove((Apartment)estate);
                    };
                    break;
                case "Room":
                    lock(_lockRooms){
                        _rooms.Remove((Room)estate);
                    };
                    break;
                case "House":
                    lock(_lockHouses){
                        _houses.Remove((House)estate);
                    };
                    break;
            }
        }

        public void PrintRandomEstateAndClones(){
            Random r = new Random();
            int i = r.Next(2);
            IEnumerable<RealEstate> estate = new List<RealEstate>();
            switch (i){
                case 0:
                    estate = Apartments;
                    break;
                case 1:
                    estate = Rooms;
                    break;
                case 2:
                    estate = Houses;
                    break;
            }
            
            if(estate.Count() > 0){
                int j = r.Next(estate.Count() - 1);
                var el = estate.ElementAt(j);
                var elClone = el.Clone();
                var elCopy = el.ShallowCopy();

                Console.WriteLine($"Original: {el}\n");
                Console.WriteLine($"Clone: {elClone}\n");
                Console.WriteLine($"Shallow copy: {elCopy}\n");

                Console.WriteLine("Change shallow copy realtor name to 'Jane Doe'...\n");
                elCopy.Realtor.NameAndSurname = "Jane Doe";
                Console.WriteLine($"Original after changing shallow copy: {el}\n");
                Console.WriteLine($"Clone: {elClone}");
            }

        }
    }
}
